Scripts/niceListAlgorithm.py

```diff
@@ -1,74 +1,97 @@
+        self.assigned_from = None        #track who is assigned to Node
+    
+    def removeEdgeFrom(self, node):      
+    #remove a name from list of incoming edges to the node
+        self.from_list.remove(node)
+    
+    def removeEdgeTo(self, node):       
+    #remove a name from the list of outgoing edges from the node
+        self.to_list.remove(node)
+        
+def permuteList(ordered_list):           
+#ensure nondeterministic outputs
+        old_list = list(ordered_list)
+        new_list = []
+        
+        for i in range(len(old_list)):
+            randIndex = random.randint(0, len(old_list) - 1)
+            new_list.append(old_list.pop(randIndex))
+         
+        return new_list
+
+def santaAssign(emails, not_allowed):
+    nodeHash = {}
+    # dictionary mapping <name> to the Node that represents <name>
+    emailsToList = permuteList(emails)
+    emailsFromList = permuteList(list(emails)) #randomize
+    for email in emails:
+        # create node, then ensure that a person is not assigned to themselves
+        nodeHash[email] = Node(emailsToList,emailsFromList)
+        nodeHash[email].removeEdgeFrom(email)
+        nodeHash[email].removeEdgeTo(email)
+        
+    for i,j in not_allowed:
+        # ensure compliance with restricted pairings
+        nodeHash[i].removeEdgeTo(j)
+        nodeHash[j].removeEdgeFrom(i)
+        
+    n = 0
+    m = 0
+    
+    while True:
+        # use depth first search to find a suitable assignment for everyone
+        activeNode = nodeHash[emails[n]]
+        if nodeHash[activeNode.to_list[m]].assigned_from == None:
+            # assign the mth person on the nth persons to_list has been assigned, if they are available
+            activeNode.assigned_to = activeNode.to_list[m]
+            nodeHash[activeNode.assigned_to].assigned_from = emails[n]
+            n += 1
+            m = 0
+            if n == len(emails):
+            # if the algorithm has assigned someone to the nth person, then a solution has been found
+                solution = []
+                for email in emails:
+                    solution.append((email,nodeHash[email].assigned_to))
+                return solution
+        elif m < len(activeNode.to_list) - 1:
+            # if the mth person on the nth person's to_list was unavailable, increment m
+            m += 1
+        else:
+            # if everyone on the nth person's to_list has been assigned, decrement n until an assignment can be made
+            while True:
+                n -= 1
+                assignment = nodeHash[emails[n]].assigned_to
+                if n == -1:
+                    # if the algorithm cycles backwards to the first person, and there is no one left on their to_list, then there is no solution
+                    return "No solution available"
+                k = nodeHash[emails[n]].to_list.index(assignment)
+                nodeHash[assignment].assigned_from = None
+                assignment = None

```

Comment:
what is this line doing? Are you trying to do the same as:
`nodeHash[emails[n]].assigned_to = None`

Reply:
I am doing that in line 77,  because
`assignment = nodeHash[emails[n]].assigned_to`
The node n was assigned to node m. My algorithm also keeps track of the fact that node m is assigned 'from' node n, so in order for my code to run properly, I need to set both node n's assigned_to variable and node m's assigned_from variable to None.
I could have done this in fewer lines of code, but I thought this was more readable. I see now that I can remove the variable k, as it is only used in line 78.

Reply:
I don't think it works like you think it does. Try running this code: https://pastebin.com/mgK02fXa

Reply:
Hmm. Are you saying that my 'assignment' variable is not a reference to the same memory allocated for
`nodeHash[emails[n]].assigned_to`
but rather just a copy?
(because something something immutable types?)
Also, this has me wondering whether or not assigning things a value of None serves any purpose...

Reply:
I don't think it has to do with immutable types. You can mutate a string. It has to due with simple types (e.g. strings, ints, doubles). They tend to be copied on assignment. And yeah your algorithm might be working fine whether you assign to null or not. But I think it's better to assign to null.

Reply:
I revised the code in this section; I created a new variable for the active Node, instead of for the string representing the assignment, since the new variable is duplicated reference, and not a copy (I think).

Reply:
looks good